### ISSUE TITLE
Node & NPM

### ISSUE THREAD

**Issue by leeoniya** [leeoniya/domvm#5]
Figure out needed polyfills? (fetch/xr, rAF, Promises, History)
Ensure everything works that's necessary under node, isomorphism.
Determine & minimize devDependencies
Qunit tests integration
Closure compiler or UglifyJS minification
Gulp? Browserify?
Publish beta

**Comment by leeoniya** [comment]
isomorphic markup generation already works on node with 0 dependencies. probably time to make a package.json and publish.

**Comment by leeoniya** [comment]
@yosbelms

I've been having problems because i don't want a lot of boilerplate for each module since i want the lib to be build-able without NPM, Browserify or Bower and still be small for concat and min (as in the README). I also want all modules to simply by included in order as script tags in the browser and work without any build process for dev. Right now I do pollute the `global` namespace in node with `domvm` and all the modules live under that...not too happy about it :(

Woul dbe nice to have something like the following but I have no idea how to dep-inject the util module so it's available to the view module uder a commonly accessible namespace. It may be possible with a ton of module boilerplate in each file, but again that would require a bunch of dependencies to build out a non-bloated concat & min file:

```js
var domvm = {
  util: require("./util"),
  view: require("./view"),
  html: require("./html"),
  watch: require("./watch"),
  route: require("./route"),
};
```

**Comment by yosbelms** [comment]
@leeoniya Regarding all requirements you mentioned before, the global pollution is the solution at first glance, at least. However I'm going deeper to see what I can do

**Comment by leeoniya** [comment]
thanks!

i skimmed through this article [1] but didnt find anything that could fully address the situation.

i would not be opposed to using some native commandline utils (head/tail/sed) to strip out any umd boilerplate/wrapper from each file prior to closure concat & compile. i just dont want a node/npm/gulp/grunt/browserify/bower to build a util meant primarily for the browser.

if npm exists and browserify is avail, then great and there can be a slightly different build setup in that case.

[1] http://bites.goodeggs.com/posts/export-this/

**Comment by yosbelms** [comment]
May be using a different build for node, or concatenating every file and removing the pollution by adding a boilerplate in the end of the bundle, it could be:
```js
// node_suffix.js
if (typeof module !== 'undefined' && typeof module.exports !== 'undefined') {
    module.exports = global.domvm;
    global.domvm   = void 0;
}
```
After that, `domvm` will be `undefined` in the global scope unlike you've explicitly required with `var domvm = require('domvm')`:

I use [shelljs](https://github.com/shelljs/shelljs) to strip/add boilerplate/wrappers, build docs, and task automation most of the time. I think using that tool + uglify is enough, [this](https://github.com/yosbelms/cor/blob/master/make) is an example of shelljs heavy usage.

**Comment by leeoniya** [comment]
shelljs looks like a nice thin build dependency under node. ~200k, ~60 files.

if you wanna whip up a script that adds the needed wrapper boilerplate and runs everything through closure, we can start with that.

**Comment by yosbelms** [comment]
Shelljs scripts are about 5k all the others are doc/test files. Anyway, the dependences are just declared in `package.json` these are part of the repo, then you can download it by running `npm install` from `./domvm` directory

**Comment by yosbelms** [comment]
I'm currently trying to run tests in several borwsers but last year borwsers doesn't support arrow functions, same with nodejs.
:(

**Comment by leeoniya** [comment]
yeah, the tests do use arrow funcs but they dont rely on them. feel free to change back to plain funcs. think i just liked the breavity and writing so many similar tests can get sore on the eyeballs. heh.

**Comment by yosbelms** [comment]
@leeoniya  It it very important to run tests in browsers with different ages. Are you planning to support just 2016 browsers? If yes, it is a weakness. :-1:

**Comment by leeoniya** [comment]
ie9+ (with polyfills):

Promise
History (router)
Fetch (mutaion observer)
rAF (redraw coalescing)
qSA (event deleg)
WeakMap (possibly)

i do not actively test in all browsers yet but i do avoid using features that cannot be polyfilled in ie9+. the tests only cover DOM/vtree correctness right now and are only the most critical 30% of the features. there's a lot more tests to write. Once that is done then we can make sure they all pass in all browsers.

then there's writing docs and demo apps and benchmark tests and packaging not to mention actually developing the lib itself. i'm the only dev so i'm focusing on a couple things at a time. help is always welcome :)

**Comment by leeoniya** [comment]
@yosbelms

> BTW, I'm really interested in this project, so I have taken the GitHub organization (https://github.com/domvm), npm package (https://www.npmjs.com/package/domvm) and the Bower package, all three ready to transfer the ownership to you, whenever you want.

I haven't seen you reply to any issues for a while and noticed you removed domvm from https://github.com/yosbelms/radar repo. If you're no longer interested in the project or don't have time to participate, let's go ahead and do these transfers now if you don't mind. It'd suck to have a disinterested party in control of the name.

Thanks!

**Comment by yosbelms** [comment]
Actually my time is short these days. However I'm still watching.

**Comment by leeoniya** [comment]
No prob, I understand. Thanks.

How does the npm ownership transfer work?

**Comment by yosbelms** [comment]
You need a npm user

**Comment by leeoniya** [comment]
https://www.npmjs.com/~leeoniya

**Comment by yosbelms** [comment]
done, check https://www.npmjs.com/~leeoniya

**Comment by leeoniya** [comment]
thank you.

Feel free to jump in and out if you have time or interest. Your input so far has been really helpful.
